Long press hint for Rich Links in Mail is solid black

rr-codes opened the pull request:
#### a010904c451cd76eea5d885b648375c79135384f
<pre>
Long press hint for Rich Links in Mail is solid black
<a href="https://bugs.webkit.org/show_bug.cgi?id=243437">https://bugs.webkit.org/show_bug.cgi?id=243437</a>
rdar://97644287

Reviewed by Aditya Keerthi and Wenson Hsieh.

Currently, a TextIndicator only surrounds the Element that is selected.
Instead, it should consider the element and all of its ancestors up until an
ancestor has a `-webkit-user-select: all` as a single unit. This provides a
better user experience and gets around the root cause of why the indicator for
the actual text is a solid color, which is due to the behavior of
`-webkit-user-select: none` with TextIndicator.

* LayoutTests/fast/text-indicator/text-indicator-nested-anchors-expected.txt: Added.
* LayoutTests/fast/text-indicator/text-indicator-nested-anchors.html: Added.
* Source/WebCore/page/TextIndicator.cpp:
(WebCore::TextIndicator::createWithNode):
* Source/WebCore/page/TextIndicator.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::textIndicatorForRange):
(WebCore::Internals::textIndicatorForNode):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performImmediateActionHitTestAtLocation):

Canonical link: <a href="https://commits.webkit.org/253097@main">https://commits.webkit.org/253097@main</a>
</pre>

whsieh commented on LayoutTests/fast/text-indicator/text-indicator-nested-anchors.html:
```diff
@@ -0,0 +1,45 @@
+            <table>
+                <tbody>
+                <tr>
+                    <td><<div>hello</div></td>
```
Nit - extra "<" before the div element here.

whsieh commented on Source/WebCore/page/TextIndicator.h:
```diff
@@ -144,6 +145,7 @@ class TextIndicator : public RefCounted<TextIndicator> {
     WEBCORE_EXPORT static Ref<TextIndicator> create(const TextIndicatorData&);
     WEBCORE_EXPORT static RefPtr<TextIndicator> createWithSelectionInFrame(Frame&, OptionSet<TextIndicatorOption>, TextIndicatorPresentationTransition, FloatSize margin = FloatSize(defaultHorizontalMargin, defaultVerticalMargin));
     WEBCORE_EXPORT static RefPtr<TextIndicator> createWithRange(const SimpleRange&, OptionSet<TextIndicatorOption>, TextIndicatorPresentationTransition, FloatSize margin = FloatSize(defaultHorizontalMargin, defaultVerticalMargin));
+    WEBCORE_EXPORT static RefPtr<TextIndicator> createWithNode(Node&, TextIndicatorPresentationTransition, FloatSize margin = FloatSize(defaultHorizontalMargin, defaultVerticalMargin));
```
Hm...is there a reason why we're generating text indicator options in `createWithNode()`, rather than taking it as an argument? (I think in general, it makes sense to let `TextIndicator` clients to specify options when creating the text indicator).

whsieh commented on LayoutTests/fast/text-indicator/text-indicator-nested-anchors-expected.txt:
```diff
@@ -0,0 +1,11 @@
+PASS successfullyParsed is true
+
+TEST COMPLETE
+<
```
Nit - still have a stray `<` in the expected file.

whsieh commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -67,6 +69,20 @@ Ref<TextIndicator> TextIndicator::create(const TextIndicatorData& data)
+
+    return createWithRange(elementRange, options, presentationTransition, margin);
+}
+
```
Maybe something like this?
```
RefPtr<TextIndicator> TextIndicator::createWithNode(Node& node, OptionSet<TextIndicatorOption> options, TextIndicatorPresentationTransition presentationTransition, FloatSize margin)
{
    Ref topMostSameLinkElement = node;

    for (auto& ancestorElement : ancestorsOfType<Element>(node)) {
        if (auto* renderer = ancestorElement.renderer(); renderer && renderer->style().effectiveUserSelect() == UserSelect::All)
            topMostSameLinkElement = ancestorElement;
    }

    return createWithRange(makeRangeSelectingNodeContents(topMostSameLinkElement.get()), options, presentationTransition, margin);
}
```
(I made several small adjustments)
- Use a smart pointer to store the top-most link on the stack, instead of a raw pointer.
- Removed an extra space before the `TextIndicatorPresentationTransition`
- Added a null check before accessing `ancestorElement.renderer()` — I don't think anything guarantees that the `renderer()` is non-null here?

pxlcoder commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -67,6 +69,20 @@ Ref<TextIndicator> TextIndicator::create(const TextIndicatorData& data)
     return adoptRef(*new TextIndicator(data));
 }
 
+RefPtr<TextIndicator> TextIndicator::createWithNode(Node& node, OptionSet<TextIndicatorOption> options,  TextIndicatorPresentationTransition presentationTransition, FloatSize margin)
```
Did you consider making this behavior a `TextIndicatorOption`, rather than a separate method?

Currently, it's not clear from this name of this method that it accounts for `user-select`, while `createWithRange` does not. There may also be other places where we do want to account for `user-select`, in which case it would be easier to add an option, than remember to use a specific method.

whsieh replied:
New `TextIndicatorOption` (as opposed to create method) sounds good to me too!

rr-codes replied:
@pxlcoder  @whsieh how would this work exactly? Since `createWithNode` takes a `Node` as its parameter, whereas `createWithRange` takes a range

whsieh replied:
I think the idea is that you would still pass in a range, but specify an option that allows `TextIndicator` to march up the DOM, to the closest `user-select: all` ancestor if needed.

(You could probably start the ancestor traversal at the common ancestor node of the given range)

pxlcoder commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -67,6 +69,20 @@ Ref<TextIndicator> TextIndicator::create(const TextIndicatorData& data)
 
+RefPtr<TextIndicator> TextIndicator::createWithNode(Node& node, OptionSet<TextIndicatorOption> options,  TextIndicatorPresentationTransition presentationTransition, FloatSize margin)
+{
+    auto topMostSameLinkElement = &node;
```
`auto*`

pxlcoder replied:
(`Ref` is even better as @whsieh commented above)

whsieh commented on Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
```diff
@@ -63,6 +63,7 @@
 #import <WebCore/DictionaryLookup.h>
 #import <WebCore/Editing.h>
 #import <WebCore/Editor.h>
+#import <WebCore/ElementAncestorIterator.h>
```
Nit - I think you can remove this `#import`.

pxlcoder commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -67,6 +69,20 @@ Ref<TextIndicator> TextIndicator::create(const TextIndicatorData& data)
 
+RefPtr<TextIndicator> TextIndicator::createWithNode(Node& node, OptionSet<TextIndicatorOption> options,  TextIndicatorPresentationTransition presentationTransition, FloatSize margin)
+{
+    auto topMostSameLinkElement = &node;
```
This variable should be renamed to be more generic, since the node is not necessarily a link (anchor) element.

whsieh commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -31,13 +31,15 @@
 #include "GeometryUtilities.h"
 #include "GraphicsContext.h"
 #include "ImageBuffer.h"
+#include "ImageOverlay.h"
```
You can also remove this `#include`

whsieh reviewed (approved):
LGTM with comments!

pxlcoder commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -69,30 +70,43 @@ Ref<TextIndicator> TextIndicator::create(const TextIndicatorData& data)
 RefPtr<TextIndicator> TextIndicator::createWithRange(const SimpleRange& range, OptionSet<TextIndicatorOption> options, TextIndicatorPresentationTransition presentationTransition, FloatSize margin)
 {
-    RefPtr frame = range.startContainer().document().frame();
+    auto rangeToUse = range;
```
Nit: `auto&`.

rr-codes replied:
I tried this, but I don't think it's possible since `range` is `const`, and then `rangeToUse` is reassigned later on, so it won't compile

pxlcoder commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -69,30 +70,43 @@ Ref<TextIndicator> TextIndicator::create(const TextIndicatorData& data)
+
+        for (auto& ancestorElement : ancestorsOfType<Element>(*rangeNode)) {
+        if (auto* renderer = ancestorElement.renderer(); renderer && renderer->style().effectiveUserSelect() == UserSelect::All)
+            indicatorNode = ancestorElement;
```
Nit: Indentation.

pxlcoder commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -69,30 +70,43 @@ Ref<TextIndicator> TextIndicator::create(const TextIndicatorData& data)
-    RefPtr frame = range.startContainer().document().frame();
+    auto rangeToUse = range;
+    if (options.contains(TextIndicatorOption::UseUserSelectAllCommonAncestor)) {
+        auto rangeNode = commonInclusiveAncestor<ComposedTree>(range);
```
Should we null-check this value? `commonInclusiveAncestor` returns a pointer, and we null-check it elsewhere.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
```diff
@@ -849,7 +849,9 @@ static void drawPDFPage(PDFDocument *pdfDocument, CFIndex pageIndex, CGContextRe
-        OptionSet<TextIndicatorOption> options { TextIndicatorOption::UseBoundingRectAndPaintAllContentForComplexRanges };
+        OptionSet<TextIndicatorOption> options { TextIndicatorOption::UseBoundingRectAndPaintAllContentForComplexRanges, TextIndicatorOption::UseUserSelectAllCommonAncestor };
+
+
```
Remove whitespace.

pxlcoder commented on Source/WebCore/page/TextIndicator.cpp:
```diff
@@ -69,30 +70,44 @@ Ref<TextIndicator> TextIndicator::create(const TextIndicatorData& data)
-    RefPtr frame = range.startContainer().document().frame();
+    auto rangeToUse = range;
+    if (options.contains(TextIndicatorOption::UseUserSelectAllCommonAncestor)) {
+        if (auto rangeNode = commonInclusiveAncestor<ComposedTree>(range); rangeNode) {
```
This can just be `if (auto* commonAncestor = commonInclusiveAncestor<ComposedTree>(range))`.

(I would also rename `rangeNode` to `commonAncestor`)

rr-codes committed:
Long press hint for Rich Links in Mail is solid black
https://bugs.webkit.org/show_bug.cgi?id=243437
rdar://97644287

Reviewed by Aditya Keerthi and Wenson Hsieh.

Currently, a TextIndicator only surrounds the Element that is selected.
Instead, it should consider the element and all of its ancestors up until an
ancestor has a `-webkit-user-select: all` as a single unit. This provides a
better user experience and gets around the root cause of why the indicator for
the actual text is a solid color, which is due to the behavior of
`-webkit-user-select: none` with TextIndicator.

* LayoutTests/fast/text-indicator/text-indicator-nested-anchors-expected.txt: Added.
* LayoutTests/fast/text-indicator/text-indicator-nested-anchors.html: Added.
* Source/WebCore/page/TextIndicator.cpp:
(WebCore::TextIndicator::createWithNode):
* Source/WebCore/page/TextIndicator.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::textIndicatorForRange):
(WebCore::Internals::textIndicatorForNode):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performImmediateActionHitTestAtLocation):

Canonical link: https://commits.webkit.org/253097@main
